src/Azure.DataApiBuilder.Mcp/Core/McpStdioServer.cs

```diff
@@ -0,0 +1,486 @@
+using Azure.DataApiBuilder.Mcp.Model;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using ModelContextProtocol.Protocol;
+
+namespace Azure.DataApiBuilder.Mcp.Core
+{
+    /// <summary>
+    /// MCP stdio server:
+    /// - Reads JSON-RPC requests (initialize, listTools, callTool) from STDIN
+    /// - Writes ONLY MCP JSON responses to STDOUT
+    /// - Writes diagnostics to STDERR (so STDOUT remains “pure MCP”)
+    /// </summary>
+    public class McpStdioServer : IMcpStdioServer
+    {
+        private readonly McpToolRegistry _toolRegistry;
+        private readonly IServiceProvider _serviceProvider;
+        private readonly string _protocolVersion;
+
+        private const int MAX_LINE_LENGTH = 1024 * 1024; // 1 MB limit for incoming JSON-RPC requests
+
+        public McpStdioServer(McpToolRegistry toolRegistry, IServiceProvider serviceProvider)
+        {
+            _toolRegistry = toolRegistry ?? throw new ArgumentNullException(nameof(toolRegistry));
+            _serviceProvider = serviceProvider ?? throw new ArgumentNullException(nameof(serviceProvider));
+
+            // Allow protocol version to be configured via IConfiguration, using centralized defaults.
+            IConfiguration? configuration = _serviceProvider.GetService<IConfiguration>();
+            _protocolVersion = McpProtocolDefaults.ResolveProtocolVersion(configuration);
+        }
+
+        /// <summary>
+        /// Runs the MCP stdio server loop, reading JSON-RPC requests from STDIN and writing MCP JSON responses to STDOUT.
+        /// </summary>
+        /// <param name="cancellationToken">Token to signal cancellation of the server loop.</param>
+        /// <returns>A task representing the asynchronous operation.</returns>
+        public async Task RunAsync(CancellationToken cancellationToken)
+        {
+            Console.Error.WriteLine("[MCP DEBUG] MCP stdio server started.");
+
+            // Use UTF-8 WITHOUT BOM
+            UTF8Encoding utf8NoBom = new(encoderShouldEmitUTF8Identifier: false);
+
+            using Stream stdin = Console.OpenStandardInput();
+            using Stream stdout = Console.OpenStandardOutput();
+            using StreamReader reader = new(stdin, utf8NoBom);
+            using StreamWriter writer = new(stdout, utf8NoBom) { AutoFlush = true };
+
+            // Redirect Console.Out to use our writer
+            Console.SetOut(writer);
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                string? line = await reader.ReadLineAsync(cancellationToken);
+                if (string.IsNullOrWhiteSpace(line))
+                {
+                    continue;
+                }
+
+                if (line.Length > MAX_LINE_LENGTH)
+                {
+                    WriteError(id: null, code: -32600, message: "Request too large");
```

Comment:
This says WriteError but still is writing to stdout in the function. Why is that? Shouldn't we write to STDERR?